lib/appium.js

```diff
@@ -344,35 +346,142 @@ class AppiumDriver extends BaseDriver {
+      if (!dstSession) {
+        throw new Error(`The session with id '${sessionId}' does not exist`);
+      }
+      // now save the response protocol given that the session driver's protocol might differ
+      protocol = dstSession.protocol;
     }
 
-    if (isAppiumDriverCommand(cmd)) {
-      return await super.executeCommand(cmd, ...args);
-    }
+    // now we define a 'handled' object which will keep track of which plugins have handled this
+    // command. we care about this because (a) multiple plugins can handle the same command, and
+    // (b) there's no guarantee that a plugin will actually call the next() method which runs the
+    // original command execution. This results in a situation where the command might be handled
+    // by some but not all plugins, or by plugin(s) but not by the default behavior. So start out
+    // this object declaring that the default handler has not been executed.
+    const handled = {default: false};
+
+    // now we define an async function which will be passed to plugins, and successively wrapped
+    // if there is more than one plugin that can handle the command. To start off with, the async
+    // function is defined as calling the default behavior, i.e., whichever of the 3 cases above is
+    // the appropriate one
+    const defaultBehavior = async () => {
+      log.info(`Executing default handling behavior for command '${cmd}'`);
+      handled.default = true; // if we make it here, we know that the default behavior is handled
+
+      if (isGetStatus) {
+        return await this.getStatus();
+      }
 
-    const sessionId = _.last(args);
-    const dstSession = await sessionsListGuard.acquire(AppiumDriver.name, () => this.sessions[sessionId]);
-    if (!dstSession) {
-      throw new Error(`The session with id '${sessionId}' does not exist`);
-    }
+      if (isUmbrellaCmd) {
+        // some commands, like deleteSession, we want to make sure to handle on *this* driver,
+        // not the platform driver
+        return await super.executeCommand(cmd, ...args);
+      }
 
-    let res = {
-      protocol: dstSession.protocol
+      // here we know that we are executing a session command, and have a valid session driver
+      return await dstSession.executeCommand(cmd, ...args);
     };
 
+    // now take our default behavior, wrap it with any number of plugin behaviors, and run it
+    const wrappedCmd = this.wrapCommandWithPlugins({cmd, args, handled, next: defaultBehavior});
+    const res = await this.executeWrappedCommand({wrappedCmd, protocol});
+
+    this.logPluginHandlerReport({cmd, handled});
+
+    return res;
+  }
+
+  wrapCommandWithPlugins ({cmd, args, next, handled}) {
+    // get any plugins which are registered as handling this command
+    const plugins = this.pluginsToHandleCmd(cmd);
+
+    plugins.length && log.info(`Plugins which can handle cmd '${cmd}': ${plugins.map((p) => p.name)}`);
```

Comment:
we should do a short return here if there are no plugins

Reply:
yeah, good thought

Reply:
actually, looking at this, what is the advantage of a short return? if the list is empty, the loop won't be executed anyway. is it just that it is more clear to you?